File: packages/android_intent_plus/lib/src/put_classes/put_char_array.dart
import 'package:android_intent_plus/src/put_classes/base/put_base.dart';

class PutCharArray extends PutBase<List<String>> {
  PutCharArray({required String key, required List<String> value})
      : super(key: key, value: value) {
    for (final char in value) {
      final length = char.length;
      if (length != 1) {
        throw RangeError.value(
          length,
          "value.length must be 1.",
        );
      }
    }
  }

  @override
  String get javaClass => 'PutCharArray';

  factory PutCharArray.fromJson({required String key, required dynamic value}) {
    return PutCharArray(key: key, value: List<String>.from(value));
  }

  @override
  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'key': key,
      'javaClass': javaClass,
      'value': value,
    };
  }
}

File: packages/android_intent_plus/android/src/main/java/dev/fluttercommunity/plus/androidintent/Bundle/PutClasses/PutCharArray.java
package dev.fluttercommunity.plus.androidintent.Bundle.PutClasses;

import android.os.Bundle;
import dev.fluttercommunity.plus.androidintent.Bundle.Constants;
import dev.fluttercommunity.plus.androidintent.Bundle.Helpers;
import dev.fluttercommunity.plus.androidintent.Bundle.PutClasses.base.PutBase;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class PutCharArray extends PutBase<List<Character>> {

  public static final String JAVA_CLASS = "PutCharArray";

  public PutCharArray(String key, List<Character> value) {
    super(key, JAVA_CLASS, value);
  }

  public static void convert(Bundle bundle, PutCharArray putCharArray) {
    bundle.putCharArray(putCharArray.key, toPrimitiveArray(putCharArray.value));
  }

  private static char[] toPrimitiveArray(final List<Character> characterList) {
    final char[] primitives = new char[characterList.size()];
    for (int i = 0; i < characterList.size(); i++) {
      primitives[i] = characterList.get(i);
    }
    return primitives;
  }

  public static PutCharArray fromJson(JSONObject jsonObject) throws JSONException {
    final ArrayList<Character> arrayList = new ArrayList<>();
    final JSONArray jsonArray = jsonObject.getJSONArray(Constants.VALUE);
    for (int i = 0; i < jsonArray.length(); i++) {
      arrayList.add(jsonArray.getString(i).charAt(0));
    }
    return new PutCharArray(jsonObject.getString(Constants.KEY), arrayList);
  }

  @Override
  public JSONObject toJson() throws JSONException {
    JSONObject jsonObject = new JSONObject();
    jsonObject.put("key", key);
    jsonObject.put("javaClass", JAVA_CLASS);
    jsonObject.put("value", new JSONArray(Helpers.convertToArrayListOfString(value)));
    return jsonObject;
  }
}
